### PR TITLE
Allow custom suggestions filter

### PR DESCRIPTION
**Pull request by rnicholus**
- [x] Fix `npm test` on recent versions of npm/node.
- [x] Add code to support (optional) custom suggestions filter.
- [x] Unit tests for new feature.
- [x] Update documentation.
- [x] Relax React-related peer dependencies to allow for 0.14.x.
- [x] Increment build minor version.
- [x] Integration test with a project that utilizes react-tags.

This will address #73, #82, and #91.

### CODE REVIEW

**Comment by prakhar1989** [comment]
👍

**Comment by stephenburgess8** [comment]
+1

**Comment by rnicholus** [comment]
This appears to be working and coded as expected. @prakhar1989 and @stephenburgess8 can you take a quick look and the diff? If all looks good, I can merge this into master, and then perhaps @prakhar1989 can `npm publish` to get react-tags-input@4.1.0 onto npm.

**Comment by prakhar1989** [comment]
Hey guys, I'm sorry I didn't get a chance to check this out. Will surely give this a go by tonight and update here accordingly. Thanks a lot again for working on this!

File: package.json [hunk 1]
```diff
@@ -43,6 +43,7 @@
     "chai": "^3.5.0",
     "enzyme": "^2.3.0",
     "jsdom": "^9.4.0",
+    "lodash": "4.14.1",
```
Comment (prakhar1989) [inline package.json]: Adding this causes the tests to break due to imports. To replicate, delete your node_modules folder and run `npm run test`. Is there a reason why this has been added?

File: README.md [hunk 2]
```diff
@@ -188,6 +189,20 @@ function(tag, currPos, newPos) {
     // remove tag from currPos and add in newPos
 }
 ```
+
+<a name="handleFilterSuggestions"></a>
+##### handleFilterSuggestions (optional)
+To assert control over the suggestions filter, you may contribute a function that is executed whenever a filtered set
+of suggestions is expected. By default, the text input value will be matched against each suggestion, and those that
+start with the entered text will be included in the filters suggestions list. If you do contribute a custom filter
```
Comment (prakhar1989) [inline README.md]: You might want to highlight **start with**  and / or link to the `startsWith` method just to make it clear to the user about how the default matching is done.

File: README.md [hunk 3]
```diff
@@ -188,6 +189,20 @@ function(tag, currPos, newPos) {
...
+function, you must return an array of suggestions. Please do not mutate the passed suggestions array.
+
+```js
+function(textInputValue, possibleSuggestionsArray) {
+    // filter suggestions and return a new suggestions array
```
Comment (prakhar1989) [inline README.md]: Can you add a small snippet to show how to use `endWith` or `contains` as an example? Few users might not perfectly understand the warning and hence code might communicate better.

File: test/reactTags-test.js [hunk 4]
```diff
@@ -52,4 +53,42 @@ describe("ReactTags", () => {
     expect($el.find('.ReactTags__tagInput input').get(0).value).to.be.empty;
 
   });
-});
+
+  describe('autocomplete/suggestions filtering', () => {
```
Comment (prakhar1989) [inline test/reactTags-test.js]: These tests are absolutely wonderful. Thanks a ton for adding these here 🚀 🚀 🚀

**Comment by prakhar1989** [comment]
Just have a few comments, rest of it looks great :)

File: package.json (hunk 1 above)
Comment (rnicholus) [inline package.json]: Which version of npm/node are you using? In node 5, I was _only_ able to get the tests to pass with this devDependency.

File: README.md (hunk 2 above)
Comment (rnicholus) [inline README.md]: Sure thing. I can do both.

File: README.md (hunk 3 above)
Comment (rnicholus) [inline README.md]: Will do! I'll probably include an example that uses `String.prototype.includes` along with a warning to use a polyfill for older browsers.

File: package.json (hunk 1 above)
Comment (prakhar1989) [inline package.json]: I'm on node `v6.1.0` and npm `v3.8.6`.

**Comment by rnicholus** [comment]
I'll make these changes tomorrow morning (CST). Shall I wait for your review before merging after I do this, or should I just merge after I'm done so that you can `npm publish` at your convenience?

File: package.json (hunk 1 above)
Comment (rnicholus) [inline package.json]: I can add an `engines` directive to package.json that directs developers to use node 5+ when building, if you'd like.

**Comment by prakhar1989** [comment]
Sure! Take your time and feel free to merge these in as and when you like. Just give me a heads-up so that I keep you blocked until the publish the module. Thanks again! 👍

**Comment by rnicholus** [comment]
Ready for `npm publish`! I would like to delay deleting this feature branch until 4.1.0 is on npm _and_ we have updated our internal test environment at Widen to point to v4.1.0 (currently we're pointing directly at this branch).

**Comment by rnicholus** [comment]
...oh, and you'll probably want to push a new git tag too

**Comment by prakhar1989** [comment]
```
react-tags git/master
❯ gp -f origin v4.1.0
Counting objects: 1, done.
Writing objects: 100% (1/1), 186 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
To https://github.com/prakhar1989/react-tags
 * [new tag]         v4.1.0 -> v4.1.0

react-tags git/master
❯ npm publish
+ react-tag-input@4.1.0
```

Enjoy! 💃

**Pull request merged**